plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+bindDat dat@(Datatype _ (TyVarDecl _ a k) _ _ _) cont =
+  bindTyName a k $
+  local (\ e -> e { geDatas = Map.insert a dat (geDatas e) }) $
+  foldr (uncurry bindTmName) cont (matchType dat : constrTypes dat)
+
+-- | Bind a binding.
+bindBind :: Binding TyName Name DefaultUni DefaultFun ()
+         -> GenTm a
+         -> GenTm a
+bindBind (DatatypeBind _ dat)              = bindDat dat
+bindBind (TermBind _ _ (VarDecl _ x ty) _) = bindTmName x ty
+-- CODE REVIEW: Should we try to generate type bindings and all the recursive types without datatypes stuff?
+-- I don't think both datatypes and this stuff should actually show up in the same code, no?
+bindBind _                                 = error "unreachable"
+
+-- | Bind multiple bindings
+bindBinds :: Foldable f => f (Binding TyName Name DefaultUni DefaultFun ()) -> GenTm a -> GenTm a
+bindBinds = flip (foldr bindBind)
+
+-- * Generators for well-kinded types
+
+-- | Give a unique "least" (intentionally vaguely specified by "shrinking order")
+-- type of that kind. Note: this function requires care and attention to not get
+-- a shrinking loop. If you think you need to mess with this function:
+-- 1. You're probably wrong, think again and
+-- 2. If you're sure you're not wrong you need to be very careful and
+--    test the shrinking to make sure you don't get in a loop.
+minimalType :: Kind () -> Type TyName DefaultUni ()
+minimalType ty =
+  case ty of
+    Type{} -> unit
+    KindArrow _ k1 k2 ->
+      case k1 : view k2 of
+        [Type{}]         -> list
+        [Type{}, Type{}] -> pair
+        _                -> TyLam () (TyName $ Name "_" (toEnum 0)) k1 $ minimalType k2
+  where
+    view (KindArrow _ k1 k2) = k1 : view k2
+    view _                   = []
+
+    unit = TyBuiltin () (SomeTypeIn DefaultUniUnit)
+    list = TyBuiltin () (SomeTypeIn DefaultUniProtoList)
+    pair = TyBuiltin () (SomeTypeIn DefaultUniProtoPair)
+
+-- | Get the types of builtins at a given kind
+builtinTys :: Kind () -> [SomeTypeIn DefaultUni]
+builtinTys Star =
+  [ SomeTypeIn DefaultUniInteger
+  , SomeTypeIn DefaultUniUnit
+  , SomeTypeIn DefaultUniBool ]
+builtinTys _ = []
+
+-- | Generate "small" types at a given kind such as builtins, bound variables, bound datatypes,
+-- and abstractions /\ t0 ... tn. T
+genAtomicType :: Kind () -> GenTm (Type TyName DefaultUni ())
+genAtomicType k = do
+  tys <- asks geTypes
+  dts <- asks geDatas
+  let atoms = [ TyVar () x | (x, k') <- Map.toList tys, k == k' ] ++
+              [ TyVar () x | (x, Datatype _ (TyVarDecl _ _ k') _ _ _) <- Map.toList dts, k == k' ]
+      builtins = map (TyBuiltin ()) $ builtinTys k
+      lam k1 k2 = do
+        x <- genMaybeFreshTyName "a"
+        TyLam () x k1 <$> bindTyName x k1 (genAtomicType k2)
+  oneofTm $ map pure (atoms ++ builtins) ++ [lam k1 k2 | KindArrow _ k1 k2 <- [k]]
+
+-- | Generate a type at a given kind
+genType :: Kind () -> GenTm (Type TyName DefaultUni ())
+genType k = onSize (min 10) $
```

Comment:
y'all need https://hackage.haskell.org/package/hedgehog-0.6.1/docs/Hedgehog-Gen.html#v:recursive

Reply:
That function suffers from exactly the problem with sizing that I hinted at earlier - trees generated with that function would - if I read the documentation correctly - generate smaller integers near the bottom of a tree. Furthermore, the choice to halve the size is a bit arbitrary - not necessarily the correct choice when dealing with wider trees. Finally, it's not clear that you want size to change uniformly in all parts of the tree - you might want a function body to be allocated more size than the argument when generating a function application - for example.